Translate cilium-etcd-service to cluster IP address without a DNS

aanm opened the pull request:
Read per commit basis.

Last commit fixes https://github.com/cilium/cilium/issues/7231

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7100)
<!-- Reviewable:end -->

aanm committed:
kvstore: specify more options when setting up a kvstore client

Without an easy way to specify more options in a KVStore client, this
commit adds the ability of doing it so by introducing a new method in
the backendModule interface.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/k8s: add GetRandomBackendIP method for services in service cache

This method will be helpful to retrieve a backend IP address of a
particular service by its ID.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: derieve etcd cluster IP from domain name

As the cilium agent retrieves the list of services from kubernetes
before starting the kvstore we can guarantee that the domain name for
each request will be translated by cilium itself.

Signed-off-by: André Martins <andre@cilium.io>

vadorovsky commented on daemon/daemon_main.go:
```diff
@@ -1101,6 +1091,42 @@ func runDaemon() {
+		goopts.DialOption = []grpc.DialOption{
+			grpc.WithDialer(func(s string, duration time.Duration) (conn net.Conn, e error) {
+				// typical service name "cilium-etcd-client.kube-system.svc"
+				fullFQDN := bytes.Split([]byte(s), []byte{'.'})
```
Why can't we just do a split operation on `s` as a string? Converting to bytes (and then to strings again) seems to be an additional effort.

aanm replied:
Yeah, you were right. I've changed this entirely.

vadorovsky reviewed (commented):
One question. The rest looks good.

tgraf commented on pkg/k8s/service_cache.go:
```diff
@@ -105,6 +105,20 @@ func NewServiceCache() ServiceCache {
+	if svc == nil {
+		return nil
+	}
+	for _, port := range svc.Ports {
```
What is protecting modification of `svc` in parallel?

aanm replied:
AFAIK `svc` is never modified.

tgraf replied:
Yes, but all other code is holding the mutex while accessing `svc`. How does one know that you made this assumption? Me changing ServiceCache in the future can't possibly know that you made this assumption in this specific part of the code. It's an absolutely minor optimization, it's already in the cache line and everything, just hold the lock until the end.

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1109,6 +1098,50 @@ func runDaemon() {
 	bootstrapStats.k8sInit.End(true)
+	// If K8s is enabled we can do the service translation automagically by
+	// looking at services from k8s and retrieve the service IP from that.
+	// This makes cilium to not depend on kube dns to interact with etcd operator
```
s/etcd operator/etcd/

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1109,6 +1098,50 @@ func runDaemon() {
+			grpc.WithDialer(func(s string, duration time.Duration) (conn net.Conn, e error) {
+				// typical service name "cilium-etcd-client.kube-system.svc"
+				idx1 := strings.IndexByte(s, '.')
+				if idx1 >= 0 {
```
Move this into a function so you can write a quick unit test for the string parsing logic

tgraf reviewed (changes requested):
Awesome, I love it.

aanm committed:
pkg/k8s: add missing documentation in ServiceCache struct

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/k8s/service_cache.go:
```diff
@@ -83,15 +83,17 @@ type ServiceEvent struct {
 type ServiceCache struct {
+	Events chan ServiceEvent
+
+	// mutex protects the maps below including the concurrent access of each
```
👍

aanm committed:
daemon: signalize when a specific k8s cache in synced with k8s

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: setup etcd connectivity in parallel of k8s cache sync

If we do not setup etcd in parallel with the k8s cache sync, Cilium can
end up in a deadlock since CNP with CIDRs require a the identity
allocator to be setup which itself requires KVStore to be setup
which prevents the cache for CNP to be synced.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/ipcache: do not block initialization of identity watcher

Since identity watcher relies in the kvstore, we can't block its
initalization otherwise we can't initialize the kvstore in parallel.

Fixes: e1ab352fa656 ("ipcache: Provide WaitForInitialSync() to wait for kvstore sync")
Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/k8s/service_cache.go:
```diff
@@ -105,6 +105,20 @@ func NewServiceCache() ServiceCache {
 }
 
+// GetBackendIP returns a random L3n4Addr that is backing the given Service ID.
+func (s *ServiceCache) GetBackendIP(svcID ServiceID) *loadbalancer.L3n4Addr {
```
Rename this to `GetRandomBackendIP()`

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1118,6 +1107,42 @@ func runDaemon() {
+				if svc != nil {
+					backendIP := d.k8sSvcCache.GetBackendIP(*svc)
+					if backendIP != nil {
+						log.Debug("BackendIP found")
```
This debug message is useless

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1118,6 +1107,42 @@ func runDaemon() {
+				} else {
+					log.Debug("Service not found")
+				}
+				log.Debugf("etcd dialing to %q", s)
```
Add to the message that the custom dialer based on service backend is used

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1118,6 +1107,42 @@ func runDaemon() {
+		goopts = &kvstore.ExtraOptions{}
+		log := log.WithField(logfields.LogSubsys, "etcd")
+		goopts.DialOption = []grpc.DialOption{
+			grpc.WithDialer(func(s string, duration time.Duration) (conn net.Conn, e error) {
```
Please add a comment that describes how this works. In particular:
 - How does retry work?
 - What happens when the service is not populated yet?
 - What is the retry interval?
 - What happens when the backend IP disappears?

tgraf commented on pkg/k8s/service.go:
```diff
@@ -113,6 +113,29 @@ func (s ServiceID) String() string {
 	return fmt.Sprintf("%s/%s", s.Namespace, s.Name)
 }
 
+// ParseServiceIDFrom returns a ServiceID derived from the given domain name.
```
s/domain name/Kubernetes service FQDN/

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1104,14 +1104,23 @@ func runDaemon() {
-	bootstrapStats.k8sInit.End(true)
+	// We need to set up etcd in parallel so we will init the k8s subsystem
+	// and wait for all caches to be synced before restoring any endpoints.
+	// This is required because CNP with CIDRs rely on kvstore connectivity
```
Please rewrite this description. It is not clear to me. How do CNP and CIDRs rely on kvstore connectivity?

tgraf commented on pkg/ipcache/kvstore.go:
```diff
@@ -382,10 +382,12 @@ var (
-		watcher = NewIPIdentityWatcher(kvstore.Client())
-		close(initialized)
-		go watcher.Watch()
+		go func() {
```
Please add the `Fixes:` tag and refer to my commit. Do we need to backport?

aanm replied:
Done

tgraf commented on pkg/ipcache/kvstore.go:
```diff
@@ -382,10 +382,12 @@ var (
-		go watcher.Watch()
+		go func() {
+			log.Info("Starting IP identity watcher")
+			watcher = NewIPIdentityWatcher(kvstore.Client())
```
This is not enough, you need a lock to protect the `watcher` variable assignment as it can now be accessed in parallel. This was the reason it was moved out.

aanm replied:
Why? The variable is never assigned in parallel as it is protected by the `initialized` channel.

tgraf reviewed (changes requested):
This is great overall. Left some minor comments.

aanm committed:
kvstore: detect if kvstore is setup with etcd-operator

This allows to run our own service translation only if
cilium-etcd-operator is setup with Cilium. Otherwise fallback to the
normal kvstore behavior where the domain name is translated in the DNS.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -269,11 +270,31 @@ func blockWaitGroupToSyncResources(waitGroup *sync.WaitGroup, informer cache.Con
+		d.k8sResourceSyncedMu.RLock()
+		c := d.k8sResourceSynced[resourceName]
+		d.k8sResourceSyncedMu.RUnlock()
+		close(c)
```
I think this close should be done before the lock is released, and then after the close, you should do `delete(d.k8sResourceSynced, resourceName)` because otherwise the same channel could be accessed and attempted to be closed by different goroutines depending on how this structure (`d.k8sResourceSynced`) is utilized in the future. 
Or, you can do : 

```
select {
case <-c:
default:
    close(c) 
}
```

As a means of checking if the channel has been closed already.

aanm replied:
Good find
